Update code to pass checks for main.

schmelz21 opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

Specifically, to resolve pre-commit failures in `main` branch, so that new pull request are successful with GitHub checks.

## 💭 Motivation and context ##

Due to the inclusion of command-line updates, and new code pertaining to `setup.py install` requirements, previous tests developed were failing. The following coincide with resolving issues https://github.com/cisagov/pe-reports/issues/7. https://github.com/cisagov/pe-reports/issues/8 and Pull Request https://github.com/cisagov/pe-reports/pull/31 - will allow for tests to pass.

Primary Updates:
- setuptools requirements
- Exception clauses to handle required assets (FileNotFoundError)
- Add code function load_customers

## 🧪 Testing ##
1. `python3 setup.py install`
2. `pytest tests/test_pe_reports.py`
3. `pre-commit run --all-files`
4. `pe-reports $(date "+%Y-%m-%d") data/ output/ --db-creds-file=cyhy_connect/`

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

Commit:
Refactor report_metrics.py eval() to ast.literal_eval

Change all instances of eval to ast.literal_eval for security reasons.

Commit:
Add '# nosec' on all occurences of subprocess

This change bypasses bandit check. See comments in issue 23

Commit:
refactor code to remedy issue#21

Commit:
refactor code to remedy issue#21

schmelz21 committed:
Updated test

schmelz21 committed:
dependancy link

schmelz21 committed:
Setup add mongodb-config

schmelz21 committed:
flake adjustment

schmelz21 committed:
Setup include git url

schmelz21 committed:
Add matplotlib to setup

schmelz21 committed:
Update Docstring

schmelz21 committed:
Delete .bandit.yml

Not part of this PR.

schmelz21 committed:
Delete .gitignore

Not part of this PR.

schmelz21 committed:
Delete report_metrics.py

Not part of this PR.

schmelz21 committed:
Delete stylesheet.py

Not part of this PR.

schmelz21 committed:
Revert "Update Docstring"

This reverts commit abafbc8a611c99ac49fbb93e6295ba884f8081c9.

schmelz21 committed:
Revert "Delete stylesheet.py"

This reverts commit 396605884f4bdca75f95f4f1e2e4c104559e6641.

schmelz21 committed:
Revert "Delete report_metrics.py"

This reverts commit 41c41a2c4e9842bd48488b36378a54536b94c239.

schmelz21 committed:
Revert "Delete .bandit.yml"

This reverts commit d37d71da6a5735ed8ad4c287ebe39203b1e2307e.

schmelz21 committed:
Revert "Delete .gitignore"

This reverts commit 49dc06383809ed32d9964cdcb3d30b8953177e1e.

schmelz21 commented:
@cduhn17, Please review. I incorporated your branch changes that were currently dealing with some of the bandit issues and specifically adjusted our tests issues. The goal is to merge into `main`, so that we can continue with your PR and future PRs to bring `main` compliant to dev-standards.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -55,6 +56,16 @@ def load_template():
+    """Export PowerPoint report set to output directory."""
+    try:
+        file = open(CUSTOMERS)
+        names_obj = json.load(file)
```
You are manually opening a file and not closing it when finished. This will handle it automatically for these purposes.

```suggestion
        with open(CUSTOMERS) as customers_file:
            names_obj = json.load(customers_file)
```

mcdonnnj commented on .bandit.yml:
```diff
@@ -12,3 +12,5 @@ tests:
 skips:
   - B101  # skip "assert used" check since assertions are required in pytests
+  - B404
+  - B603
```
These should be individually `nosec`'d after review instead of just ignoring these checks completely.

```suggestion
```

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -76,9 +87,9 @@ def convert_to(folder, source, timeout=None):
-    process = subprocess.run(
-        args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, timeout=timeout
-    )
+    process = subprocess.run(  # nosec
+        args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, timeout=timeout  # nosec
+    )  # nosec
```
I believe this is all that's strictly necessary.

```suggestion
    process = subprocess.run(
        args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, timeout=timeout  # nosec
    )
```

cduhn17 replied:
@mcdonnnj , @schmelz-ctr 

I updated my local code per this suggestion and flake8 fails. I believe that for the tests to pass the # nosec has to remain in the code.

schmelz21 replied:
@mcdonnnj, @cduhn17 - I can confirm, between the flake8 reformatting of the long line and bandit's warning of `subprocess` it appears the interpretation `#nosec` needs to present on each line. 

Can I suggest we keep the original code to pass bandit errors for now and to keep in scope of PR, and create a TODO and new issue addressing bandit and subprocess in this instance.

mcdonnnj replied:
@schmelz-ctr Based on the arguments you are passing this call to `subprocess.run()` should have the `shell=False` option. Once that option is added the recommendation I made should hold true. If `black` breaks up the input arguments then the line with `shell=False` is the one that will need the nosec directive. Please see [here](https://github.com/cisagov/action-lineage/blob/27eddfe4d87f0aba5f27b3efcbbdcfef5c8c3bc0/src/lineage/entrypoint.py#L48-L54) as a reference example.

schmelz21 replied:
looking - thanks

schmelz21 replied:
That was it.. Resolved ae9a22f

cduhn17 replied:
@mcdonnnj , 

Looking for your approval that your request has been satisfied. See commit [ae9a22f](https://github.com/cisagov/pe-reports/commit/ae9a22fea9a0044b62a0e173a658302110bd3e66). Thanks

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +359,8 @@ def generate_reports(db, datestring, data_directory, output_directory):
-            subprocess.call(["mkdir", f"{output_directory}/ppt"])
-            subprocess.call(["mkdir", f"{output_directory}/{_id}"])
+            subprocess.call(["mkdir", f"{output_directory}/ppt"])  # nosec
+            subprocess.call(["mkdir", f"{output_directory}/{_id}"])  # nosec
```
This should be handled [using the `os` library](https://docs.python.org/3/library/os.html#os.makedirs). Please only use `subprocess` as a last resort.

cduhn17 replied:
Made [commit](https://github.com/cisagov/pe-reports/pull/33#:~:text=Verified-,9f154c6,-Update%20.gitignore) and [commit](https://github.com/cisagov/pe-reports/pull/33#:~:text=Verified-,d416eb2,-Update%20src/pe_reports), that refactors the code to use "os" in place of "subprocess"

schmelz21 replied:
Looks to be resolved 7b81319

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -521,18 +527,17 @@ def main():
 
     # Create output directory
-    subprocess.call(["mkdir", args["OUTPUT_DIRECTORY"]])
+    subprocess.call(["mkdir", args["OUTPUT_DIRECTORY"]])  # nosec
```
Same comment about the `os` library.

cduhn17 replied:
@mcdonnnj , 

This has been update per commit [d416eb2](https://github.com/cisagov/pe-reports/pull/33/commits/d416eb28715959d3311f292332dee93cdb862ca1)

mcdonnnj commented on .gitignore:
```diff
@@ -9,3 +9,4 @@ __pycache__
 .pytest_cache
 .python-version
 *.egg-info
+argValidation.py
```
This is something that should go in a local configuration.

```suggestion
```

mcdonnnj commented on tests/test_pe_reports.py:
```diff
@@ -65,6 +65,7 @@ def test_log_levels(level):
             "2021-01-01",
             "input/",
             "output/",
+            "--db-creds-file=input/",
```
This is not a path to a file.

cduhn17 replied:
@mcdonnnj , 

The commit [e6174c3](https://github.com/cisagov/pe-reports/pull/33#:~:text=Verified-,e6174c3,-Merge%20state) address your comment.

schmelz21 replied:
Adjusted on due to comment https://github.com/cisagov/pe-reports/pull/33#discussion_r694236396 and resolved in b980e9e

mcdonnnj commented on tests/test_pe_reports.py:
```diff
@@ -89,6 +90,7 @@ def test_bad_log_level():
             "2021-01-01",
             "input/",
             "output/",
+            "--db-creds-file=input/",
```
This is not a path to a file.

cduhn17 replied:
@mcdonnnj , 

This has been addresses per commit [3d0ef88](https://github.com/cisagov/pe-reports/pull/33/commits/3d0ef88e9f8376e173731191a6861096924cd133)

schmelz21 replied:
Adjusted on due to comment https://github.com/cisagov/pe-reports/pull/33#discussion_r694236396 and resolved in b980e9e

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -521,18 +527,17 @@ def main():
+        logging.error("%s : Missing input data. No report generated.", not_found)
         print("")
-        return 1
+        return 0
```
`0` should only be returned in an error-free state.

```suggestion
        return 1
```

schmelz21 replied:
0bb568a

mcdonnnj replied:
That does not fix the problem here. This should still return a `1` to indicate an error state.

mcdonnnj reviewed (changes requested):
Thanks for the work! Please see the changes from my first pass over this PR.

cduhn17 committed:
Update src/pe_reports/report_generator.py

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

cduhn17 committed:
Update .bandit.yml

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -521,18 +527,17 @@ def main():
 
     # Create output directory
-    subprocess.call(["mkdir", args["OUTPUT_DIRECTORY"]])
+    subprocess.call(["mkdir", args["OUTPUT_DIRECTORY"]])  # nosec
```
```suggestion
    os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
```

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +359,8 @@ def generate_reports(db, datestring, data_directory, output_directory):
             # Create folders in output directory
-            subprocess.call(["mkdir", f"{output_directory}/ppt"])
-            subprocess.call(["mkdir", f"{output_directory}/{_id}"])
+            subprocess.call(["mkdir", f"{output_directory}/ppt"])  # nosec
```
```suggestion
            os.mkdir(f"{output_directory}/ppt")
```

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +359,8 @@ def generate_reports(db, datestring, data_directory, output_directory):
-            subprocess.call(["mkdir", f"{output_directory}/ppt"])
-            subprocess.call(["mkdir", f"{output_directory}/{_id}"])
+            subprocess.call(["mkdir", f"{output_directory}/ppt"])  # nosec
+            subprocess.call(["mkdir", f"{output_directory}/{_id}"])  # nosec
```
```suggestion
            os.mkdir(f"{output_directory}/_id")
```

cduhn17 committed:
Update src/pe_reports/report_generator.py

refactor code to use os.mkdir dir in place of subprocess.call['mkdir'

cduhn17 committed:
Update src/pe_reports/report_generator.py

refactor code to use os.mkdir dir in place of subprocess.call['mkdir'

cduhn17 committed:
Update .gitignore

remove argValidation.py

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

cduhn17 committed:
Update src/pe_reports/report_generator.py

refactor code to use os.mkdir dir in place of subprocess

dav3r commented:
@cduhn17 When you accept [commit suggestions in the GitHub web GUI](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request), I strongly encourage you to replace the default commit message title (e.g. "Update src/pe_reports/report_generator.py") with something more specific (e.g. "Use os.mkdir in place of subprocess.call").  If necessary, feel free to add additional context in the optional "extended description" field.

In addition, we hope that you always follow the [seven rules of a great commit message](https://chris.beams.io/posts/git-commit/#seven-rules) (as mentioned in our [cisagov development guide](https://github.com/cisagov/development-guide)).

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -521,18 +527,17 @@ def main():
 
     # Create output directory
-    subprocess.call(["mkdir", args["OUTPUT_DIRECTORY"]])
+    os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
```
try:
   os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
except FileExistsError as err:
    logging.error(f"The output directory exists {err}", exc_info=True)
        return 0

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +359,8 @@ def generate_reports(db, datestring, data_directory, output_directory):
             # Create folders in output directory
-            subprocess.call(["mkdir", f"{output_directory}/ppt"])
-            subprocess.call(["mkdir", f"{output_directory}/{_id}"])
+            os.mkdir(f"{output_directory}/ppt")
```
```suggestion
            try:
                os.mkdir(f"{output_directory}/ppt")
                os.mkdir(f"{output_directory}/_id")
            except FileExistsError as err:
                logging.error(f"The output directory exists {err}", exc_info=True)
                return 0
```

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +359,8 @@ def generate_reports(db, datestring, data_directory, output_directory):
 
             # Create folders in output directory
-            subprocess.call(["mkdir", f"{output_directory}/ppt"])
-            subprocess.call(["mkdir", f"{output_directory}/{_id}"])
```
try:
                os.mkdir(f"{output_directory}/_id")
            except FileExistsError as err:
                logging.error(f"The output directory exists {err}", exc_info=True)
                return 0

cduhn17 committed:
Update src/pe_reports/report_generator.py

refactor code to check for existence of directory and return 0 if directory exists.

cduhn17 committed:
Update src/pe_reports/report_generator.py

refactor code to remove unnecessary # nosec

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -521,18 +532,17 @@ def main():
 
     # Create output directory
-    subprocess.call(["mkdir", args["OUTPUT_DIRECTORY"]])
+    os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
```
```suggestion
    try:
        os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
    except FileExistsError as err:
        logging.error(f"The output directory exists {err}", exc_info=True)
        return 0
```

cduhn17 commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +359,13 @@ def generate_reports(db, datestring, data_directory, output_directory):
+            except FileExistsError as err:
+                logging.error(f"The output directory exists {err}", exc_info=True)
+                return 0
+            os.mkdir(f"{output_directory}/_id")
```
```suggestion
            
```

cduhn17 committed:
Update src/pe_reports/report_generator.py

remove code not used.

cduhn17 committed:
Add output directory existence validation

Add try block to validate existence of output directory.

cduhn17 committed:
Update database credential directory

Removed an invalid directory for the the database credential .yaml file and replaced with the default name and location as stated in Usage.

cduhn17 committed:
Update directory for database credential file

The directory call out has been changed to reflect what is detailed as the default directory for the database credentials.

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -65,6 +65,7 @@ def test_log_levels(level):
             "2021-01-01",
             "input/",
             "output/",
+            "--db-creds-file=/secrests/database_creds.yml",
```
Is this a typo?  Did you mean `--db-creds-file=/secrets/database_creds.yml`?

More importantly, why are you passing in a `--db-creds-file` at all if it is [an optional parameter](https://github.com/cisagov/pe-reports/blob/3d0ef88e9f8376e173731191a6861096924cd133/src/pe_reports/report_generator.py#L11-L13) and it isn't relevant to this test case?

schmelz21 replied:
@dav3r the `--db-creds-file` is a requirement to get actual report output, and if not present the default arg- `/secrets/database_creds.yml` loads - hence optional. 

If a `database_creds.yml` file is not present the code handles the error in an exception and the module will run to completion. So I see what you mean... if it is not necessary to pass the test, so why add it. Originally, I thought it was good to be included due to continuity of other test args that were included. It will be removed.

schmelz21 replied:
Resolved in b980e9e

dav3r replied:
Thanks @schmelz-ctr 👍

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -89,6 +90,7 @@ def test_bad_log_level():
             "2021-01-01",
             "input/",
             "output/",
+            "--db-creds-file=/secrests/database_creds.yml",
```
Same question as above:  Is there any value in adding `--db-creds-file` here?  It seems to me like there is not.

schmelz21 replied:
Resolved in b980e9e

cduhn17 committed:
Update return value

Refactor code to reflect a failure at the return value(i.e. 1), in place of a successful return value (i.e. 0).

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

schmelz21 committed:
Removed test arg yml

schmelz21 committed:
Add shell arg to subprocess

schmelz21 committed:
Remove Except Traceback

schmelz21 committed:
Removed TODO Add command line functionality. #8

schmelz21 commented:
@mcdonnnj, @cduhn17, @dav3r, 
Please see updates. 
Note we are trying to limit scope to the "tests" and "setuptools" issues  for `main` so that we can get the PE active code (PE 0.1.1 - local) up to date to dev-standards. `develop` is reserved for PE reports 1.0.0. 

This PR was to resolve testing issues so that https://github.com/cisagov/pe-reports/pull/31 is not blocked.  
Sub-sequential PRs will follow addressing Issues according.

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -26,7 +26,7 @@
 import os
 import re
-import subprocess
+import subprocess  # nosec
```
An additional comment should be added here explaining why subprocess is needed here and why `nosec` is used.  For a good example, see [here](https://github.com/cisagov/openvpn-server-tf-module/blob/249144b14eb80af7616f033308d3aa274e85c294/cloudinit/link-nessus-agent.py#L11-L14).

cduhn17 replied:
@dav3r 
Comments added per recommendations. See commit [d80d8d9](https://github.com/cisagov/pe-reports/pull/33/commits/d80d8d9c559a34d1f987e2c32cc409e777500e80)

dav3r replied:
You appear to have copy/pasted [this code comment](https://github.com/cisagov/openvpn-server-tf-module/blob/249144b14eb80af7616f033308d3aa274e85c294/cloudinit/link-nessus-agent.py#L78-L83), which is not particularly useful.  For one thing, the error that bandit triggers here is B404, not B603.  In addition, the explanation that you copied/pasted makes no sense here since it doesn't explain how you are using `subprocess` safely in this file or why you cannot use a safer alternative.

cduhn17 replied:
@dav3r , 

I made changes to the comment referring to the bandit code and added reasons why we are using subprocess safely.   See commit [2cb2cce](https://github.com/cisagov/pe-reports/commit/2cb2cce4b3db92fb9516134870ac90d3234e270e)

dav3r replied:
In 2cb2cce, you wrote:
```
# Bandit triggers B404 here, but we're using subprocess.run()
# safely here, since the variable only provides mechanism for stdout and stderr.
```

I don't understand what you're saying in this sentence or what variable you are referring to here.  Can you please clarify this comment?

cduhn17 replied:
@dav3r , 

I made some modifications to the comments see commit [0609e49](https://github.com/cisagov/pe-reports/pull/33/commits/0609e49cd9bb745e266ba6d677cf8de9ebe06889) to help clarify. What I think to be the problem with using subprocess is that this module can introduce a way inject code from a shell.  The changes that I made to the comments now state the variable name where subprocess is used and that the results are output only. I also changed the bandit codes that you pointed out in your last comment.

dav3r replied:
Thanks, that is an improvement.  You should also reflow your comment lines so that they are all shorter than 80 characters (excluding the line with the bandit URL), e.g.:
```
# Bandit triggers B404 here, but we're using subprocess.run().  The action to
# import subprocess is safe as there is no function call here that would
# provide a way to inject an operation that would produce hazardous results.
# For more details on B404, see here:
# https://bandit.readthedocs.io/en/latest/blacklists/blacklist_imports.html#b404-import-subprocess
```

cduhn17 replied:
@dav3r 
I will do that.

dav3r replied:
While you shortened the lines in https://github.com/cisagov/pe-reports/pull/33/commits/131872589a46b7cb4d4118668a98ce2e8683a128, I still count four lines in these two comments that are longer than 80 characters.

dav3r replied:
I think this comment looks good now. 👍

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -76,9 +87,15 @@ def convert_to(folder, source, timeout=None):
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        timeout=timeout,
+        shell=False,  # nosec
```
Please add a comment here explaining the `nosec`, like we do [here](https://github.com/cisagov/openvpn-server-tf-module/blob/249144b14eb80af7616f033308d3aa274e85c294/cloudinit/link-nessus-agent.py#L78-L83).

cduhn17 replied:
@dav3r 
Comments added per recommendations. See commit [d80d8d9](https://github.com/cisagov/pe-reports/pull/33/commits/d80d8d9c559a34d1f987e2c32cc409e777500e80)

dav3r replied:
You appear to have copy/pasted the [code comment I mentioned](https://github.com/cisagov/openvpn-server-tf-module/blob/249144b14eb80af7616f033308d3aa274e85c294/cloudinit/link-nessus-agent.py#L78-L83) as is, without modifying the explanation of why your input (`args`) are safe to pass into `subprocess` - please correct this.

cduhn17 replied:
@dav3r , 

I made changes to the comment referring to the bandit code and added reasons why we are using subprocess safely.   See commit [2cb2cce](https://github.com/cisagov/pe-reports/commit/2cb2cce4b3db92fb9516134870ac90d3234e270e)

dav3r replied:
In https://github.com/cisagov/pe-reports/pull/33/commits/2cb2cce4b3db92fb9516134870ac90d3234e270e, you incorrectly changed `B603` to `B404` for this comment.  B404 is correct for the first `nosec` near the top of the file, but this `nosec` really is a `B603`.  

As with the first `nosec` comment, this comment is very confusing to me- please clarify it as well.

dav3r replied:
This comment still has some problems, as of https://github.com/cisagov/pe-reports/commit/131872589a46b7cb4d4118668a98ce2e8683a128:
1) I think you are misunderstanding what `B603` is concerned about, namely that the `args` input variable to `subprocess.run` is sanitized and secure.  Since that value is [hard-coded](https://github.com/cisagov/pe-reports/blob/131872589a46b7cb4d4118668a98ce2e8683a128/src/pe_reports/report_generator.py#L88-L96), the risks mentioned by `B603` are minimal here.
2) There is a trailing whitespace character at the end of the line that ends with `the local OS `.
3) There are two misspelled words: `posibility`, `commnication`
4) There are several lines longer than 80 characters (as mentioned in the other conversation).

cduhn17 replied:
@dav3r 
In item 1 from your last comment do we want the comment to reflect the verbiage from item 1?
I have items2-4 completed per commit [241f50b](https://github.com/cisagov/pe-reports/pull/33/commits/241f50b895f8feb4039911225adb70e864f59f9d)

dav3r replied:
Yes @cduhn17, please update the comment here so that it accurately explains why it is safe to `nosec` the `subprocess.run` command.  Feel free to use some of my verbiage or write your own.

dav3r replied:
I think this comment looks good now. 👍

dav3r commented:
@schmelz-ctr can you please update the title, description and other text at the top of this PR to ensure that it is still accurate?  At minimum, I don't see updates to any tests in this PR so the title and this text are incorrect:
> Update to tests - test_log_levels, test_bad_log_level

cduhn17 committed:
Add comments justifying # nosec usage

A comment has been added clarifying the use of # nosec.

schmelz21 committed:
linting issue

schmelz21 commented:
> @schmelz-ctr can you please update the title, description and other text at the top of this PR to ensure that it is still accurate? At minimum, I don't see updates to any tests in this PR so the title and this text are incorrect:
> 
> > Update to tests - test_log_levels, test_bad_log_level

@dav3r ,
Title has been updated and references to tests have been remove. Please review.

dav3r commented:
> Title has been updated and references to tests have been remove. Please review.

That looks better now, thanks!

dav3r reviewed (changes requested):
I can re-review after those `nosec` comments are cleaned up.

cduhn17 committed:
Update bandit comments 

Update changes the bandit notification code to use correct code and add relevant reasons why it is safe to use subprocess in this instance.

cduhn17 committed:
Update bandit comments for claification

Update bandit comments at the import subprocess call to clarify why it is safe to use subprocess, bandit comments code to 603 related to the use of subprocess at the 'process' variable followed by clarification as to why it is safe to use subprocess.

cduhn17 committed:
Update bandit comment line length 

Shortened all bandit comment lines

Commit:
Fix lint issue

dav3r commented:
Side note that occurred to me while responding to a conversation above: What happens in [this code](https://github.com/cisagov/pe-reports/blob/131872589a46b7cb4d4118668a98ce2e8683a128/src/pe_reports/report_generator.py#L86-L115) if [`sys.platform == "darwin"`](https://github.com/cisagov/pe-reports/blob/131872589a46b7cb4d4118668a98ce2e8683a128/src/pe_reports/report_generator.py#L118-L121) is false?  That should be tested and if it causes an error, an issue should be created to document it.

Commit:
Fix spelling,whitespace and line length

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -26,7 +26,15 @@
+# that would
+# provide a way to inject an operation that would produce hazardous results.
+# For more details on B404 see
+# here:
```
Reflow comment text in a more visually appealing way and add a word in that was accidentally deleted.
```suggestion
# Bandit triggers B404 here, but we're using subprocess.run() safely.
# The action to import subprocess is safe as there is no function call here
# that would provide a way to inject an operation that would produce 
# hazardous results.  For more details on B404 see here:
```

cduhn17 committed:
Reflow comment text

Format text to be more visually appealing.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

Commit:
Change bandit comment

This bandit comment change defines better reasons why its safe to use subprocess in this insteance.

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -26,7 +26,13 @@
+
+# Bandit triggers B404 here, but we're using subprocess.run() safely.
+# The action to import subprocess is safe as there is no function call here
+# that would provide a way to inject an operation that would produce 
```
Remove whitespace inadvertently left in during previous commit suggestion (this is causing [a linting failure](https://github.com/cisagov/pe-reports/runs/3446258487)).
```suggestion
# that would provide a way to inject an operation that would produce
```

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -76,9 +93,21 @@ def convert_to(folder, source, timeout=None):
+    # Bandit triggers B603 here, but we're using subprocess.run()
+    # safely here, since the args input variable to subprocess.run is sanitized
+    # and secure. Since that value is hard-coded, the risks mentioned by B603
+    # are minimal For more details on B603 see here:
```
Add a missing period.
```suggestion
    # are minimal. For more details on B603 see here:
```

dav3r reviewed (approved):
Thanks for addressing all of my concerns! 🚀

schmelz21 commented:
@mcdonnnj are you able to schedule a re-review. Thanks.

schmelz21 commented:
> Thanks for the work! Please see the changes from my first pass over this PR.

@mcdonnnj are we able to resolve this conversation. Thanks.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -55,6 +62,16 @@ def load_template():
+            names_obj = json.load(customers_file)
+    except FileNotFoundError as not_found:
+        logging.error("%s : Missing input data. No report generated.", not_found)
+    return names_obj
```
This should be initialized to an empty dictionary before the `try` block to cleanly handle the peculiar case of the file missing.

cduhn17 replied:
@mcdonnnj , 

Implemented recommendation per commit [a29600a](https://github.com/cisagov/pe-reports/pull/33/commits/a29600a58b1bc97062263a84399fa18a45a52cd8)

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -295,12 +324,7 @@ def generate_reports(db, datestring, data_directory, output_directory):
-        "/Users/schmelzs/Repos/PE_Report_Scripts/pe-reports_v1.0/pe-reports/src/pe_reports/org_names.json"
-    )
-    names_obj = json.load(f)
+    names_obj = load_customers()
```
If this is an empty dictionary you should bail out since that means organization data couldn't be loaded.

cduhn17 replied:
@mcdonnnj , 

I made an adjustment per commit [cb45751](https://github.com/cisagov/pe-reports/pull/33/commits/cb45751df9557cabc8534c2ca9bc79861ab97cf9) that addresses the need to verify the org_names.json file is present and has data in the file.

mcdonnnj replied:
There is still no check that there is a problem with the returned value from `load_customers()`. If there is a problem accessing `org_names.json` you currently output a log message, return a `None`, and then in https://github.com/cisagov/pe-reports/blob/8f982085d507fa59b328347175899225f57153ea/src/pe_reports/report_generator.py#L333
you immediately try and access the result without verifying it is a non-empty dictionary.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -353,8 +377,12 @@ def generate_reports(db, datestring, data_directory, output_directory):
+                os.mkdir(f"{output_directory}/ppt")
+                os.mkdir(f"{output_directory}/_id")
+            except FileExistsError as err:
+                logging.error(f"The output directory exists {err}", exc_info=True)
+                return 0
```
Is existing output directories a fail state? If so it should return a non-zero value to indicate failure.

cduhn17 replied:
@mcdonnnj , 

I have updated the code to better handle the error state per commit [0bb568a](https://github.com/cisagov/pe-reports/pull/33/commits/0bb568a663eb003c82d860ba857afaf1c140f77d)

dav3r replied:
Since I was confused here, I wanted to point out that 0bb568a does not address the non-zero return value, but you did fix this in https://github.com/cisagov/pe-reports/pull/33/commits/cb45751df9557cabc8534c2ca9bc79861ab97cf9.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -510,29 +538,26 @@ def main():
+    try:
+        os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
+    except FileExistsError as err:
+        logging.error(f"The output directory cannot be created. {err}")
+        return 0
```
Same question about existing output directories with the addition that the error message does not accurately describe the exception.

cduhn17 replied:
@mcdonnnj , 

I have update the code to better describe the error state per commit [0bb568a](https://github.com/cisagov/pe-reports/pull/33/commits/0bb568a663eb003c82d860ba857afaf1c140f77d)

mcdonnnj replied:
Those changes need to be made here as well. This logging message still does not accurately reflect the exception you are catching and it should return `1` to indicate an error.

mcdonnnj reviewed (changes requested):
Thank you for your work on this PR. I had some suggestions and some questions about fail states I would like explained a bit more.

Commit:
Add emtpy dictionary to names_obj and clean up whitespace

Commit:
Update code with proper error state

Commit:
Add validation presence of  data file

Commit:
Update return value at load_customers

Commit:
Update code formatting issue

dav3r commented on src/pe_reports/report_generator.py:
```diff
@@ -55,6 +62,19 @@ def load_template():
+def load_customers():
+    """Export PowerPoint report set to output directory."""
+    try:
+        names_obj = {}
```
We typically prefer this syntax for clarity.

```suggestion
        names_obj = dict()
```

schmelz21 replied:
There is an LGTM alert with this addition. Both as `{} `or `dict()`
We can suppress with `lgtm [py/multiple-definition]` at end of line 68. Would this be acceptable for dev-standards.

jsf9k replied:
The alert has nothing to do with the syntax.  The problem is that `names_obj` is defined here as well as on line 71.  Just delete line 68 and use my suggestion below.

dav3r replied:
Doh - that makes sense - suggestion withdrawn!

cduhn17 commented:
@mcdonnnj, @schmelz-ctr 
Please see the updates and comments to close unresolved conversations. Thanks

schmelz21 committed:
Update src/pe_reports/report_generator.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k commented on src/pe_reports/report_generator.py:
```diff
@@ -55,6 +62,19 @@ def load_template():
+    try:
+        names_obj = {}
+        if os.path.getsize(CUSTOMERS) != 0 and os.path.exists(CUSTOMERS):
+            with open(CUSTOMERS) as customers_file:
+                names_obj = json.load(customers_file)
+            return names_obj
```
There is no need to define a variable for `names_obj` since it is only used to return a value.
```suggestion
            with open(CUSTOMERS) as customers_file:
                return json.load(customers_file)
```

jsf9k replied:
Don't forget to delete line 68 as well.

jsf9k replied:
Line 68 is still there.

cduhn17 committed:
Remove names_obj variable and replace return value

The variable names_obj is removed and the value of the names_obj variable will be the new return value.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

Commit:
Remove names_obj variable

schmelz21 commented:
Hi @mcdonnnj , are we able to resolve conversations and your request change. Thanks.

stewartl97 commented:
@mcdonnnj Have the changes that were submitted satisfactorily resolve the issues you identified?

mcdonnnj commented:
> 
> 
> @mcdonnnj Have the changes that were submitted satisfactorily resolve the issues you identified?

Sorry for the delay. This is on my list to finish re-review on today.

schmelz21 commented:
> > @mcdonnnj Have the changes that were submitted satisfactorily resolve the issues you identified?
> 
> Sorry for the delay. This is on my list to finish re-review on today.

Ok.. I was just closing outdated and made the issue. Wasn't sure if we were good to go. 
I will hold off then.. Thanks for jumping in with the heads up and your planed re-review.

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -55,6 +62,18 @@ def load_template():
+                return json.load(customers_file)
+    except FileNotFoundError as not_found:
+        logging.error("%s : Missing input data. No report generated.", not_found)
+        return
```
This keeps the return type consistent.

```suggestion
        return dict()
```

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -55,6 +62,18 @@ def load_template():
 
+def load_customers():
+    """Export PowerPoint report set to output directory."""
+    try:
+
+        if os.path.getsize(CUSTOMERS) != 0 and os.path.exists(CUSTOMERS):
```
If you're explicitly checking that a file exists then it's the first thing that should occur.

```suggestion
    try:
        if os.path.exists(CUSTOMERS) and os.path.getsize(CUSTOMERS) != 0:
```

mcdonnnj commented on src/pe_reports/report_generator.py:
```diff
@@ -352,9 +378,24 @@ def generate_reports(db, datestring, data_directory, output_directory):
-            subprocess.call(["mkdir", f"{output_directory}/ppt"])
-            subprocess.call(["mkdir", f"{output_directory}/{_id}"])
+            # Create folders in output directory if folders dont exists. If the folders exists remove them and create new directory.
+            try:
+                if not os.path.exists(f"{output_directory}/ppt") or not os.path.exists(
+                    f"{output_directory}/_id"
+                ):
+                    os.mkdir(f"{output_directory}/ppt")
+                    os.mkdir(f"{output_directory}/_id")
+                else:
+                    os.remove(f"{output_directory}/ppt")
+                    os.remove(f"{output_directory}/_id")
+                    os.mkdir(f"{output_directory}/ppt")
+                    os.mkdir(f"{output_directory}/_id")
```
This should uncomplicate the logic here.

```suggestion
            try:
                for subdir in ("_id", "ppt"):
                    if os.path.exists(f"{output_directory}/{subdir}"):
                        os.remove(f"{output_directory}/{subdir}")
                    os.mkdir(f"{output_directory}/{subdir}")
```

mcdonnnj reviewed (changes requested):
Thank you for your continued work. I have some suggestions based on other changes you've made. Additionally there are still unresolved elements of my last review that I have commented on again.

schmelz21 committed:
load_customers() consistent return value

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

schmelz21 committed:
load_customers() Re-prioritize the check for file

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

schmelz21 committed:
Simplify directory check creation

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

Commit:
Update code with proper error state

schmelz21 committed:
Revert "Update code with proper error state"

This reverts commit 9d12e9172046beb5d1fb27080e4190ce4f47e92e.

Commit:
Check agencies not empty and clarify error state

cduhn17 commented:
@mcdonnnj , @schmelz-ctr 
Per commit[ 7115dd4](https://github.com/cisagov/pe-reports/pull/33/commits/7115dd4ff176abfaf8144065cd03f5b35d74ff36) has addressed your [comments](https://github.com/cisagov/pe-reports/pull/33#discussion_r711311602) about making sure agencies are present and a more clearly stated error state for when a directory exists. The thought was that its not exactly an error so the logging has been changed to INFO and that it is acceptable for the directory to exists.  Please review and provide guidance for changes.

schmelz21 commented:
> @mcdonnnj , @schmelz-ctr
> Per commit[ 7115dd4](https://github.com/cisagov/pe-reports/pull/33/commits/7115dd4ff176abfaf8144065cd03f5b35d74ff36) has addressed your [comments](https://github.com/cisagov/pe-reports/pull/33#discussion_r711311602) about making sure agencies are present and a more clearly stated error state for when a directory exists. The thought was that its not exactly an error so the logging has been changed to INFO and that it is acceptable for the directory to exists. Please review and provide guidance for changes.

@cduhn17 if to imply it is not an error but more of a check, can we then remove traceback arg `exc_info=True`. Alternatively we could possibly use` logging.warning` too if it satisfies @mcdonnnj concerns.

Code:
```
    try:
        os.mkdir(f"{args['OUTPUT_DIRECTORY']}")
    except FileExistsError as err:
        logging.info(f"The output directory exists. {err}")
        return 0
```

stewartl97 commented:
As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE.  Is there any possibility we can have them completed before I have to redirect the team to continue other work?

schmelz21 commented:
> As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?

@mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. 
Are you able to re-review @cduhn17 's last commit, inputs. Thanks

mcdonnnj commented:
> 
> 
> > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> 
> @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks

@schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.

https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332

schmelz21 commented:
> > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > 
> > 
> > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> 
> @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> 
> https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332



> > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > 
> > 
> > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> 
> @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> 
> https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332

@mcdonnnj, appreciate the quick response and apologies for the miss... circling in @cduhn17 to review and will get back with updates.

cduhn17 commented:
> > > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > > 
> > > 
> > > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> > 
> > 
> > @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332
> 
> > > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > > 
> > > 
> > > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> > 
> > 
> > @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332
> 
> @mcdonnnj, appreciate the quick response and apologies for the miss... circling in @cduhn17 to review and will get back with updates.

@schmelz21 , @mcdonnnj 
Per commit [eeb47ed](https://github.com/cisagov/pe-reports/pull/33/commits/eeb47ed69c0e211942434dde817043e0be412267) I was thinking we had addressed the question of if the load_customer return value is empty or not right in the load_customer function. https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L65-L73

mcdonnnj commented:
> 
> 
> > > > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > > > 
> > > > 
> > > > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> > > 
> > > 
> > > @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> > > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332
> > 
> > 
> > > > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > > > 
> > > > 
> > > > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> > > 
> > > 
> > > @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> > > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332
> > 
> > 
> > @mcdonnnj, appreciate the quick response and apologies for the miss... circling in @cduhn17 to review and will get back with updates.
> 
> @schmelz21 , @mcdonnnj Per commit [eeb47ed](https://github.com/cisagov/pe-reports/pull/33/commits/eeb47ed69c0e211942434dde817043e0be412267) I was thinking we had addressed the question of if the load_customer return value is empty or not right in the load_customer function.
> 
> https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L65-L73

@cduhn17 All that has been done is making sure that the function returns a dictionary no matter what. However if there is an issue accessing the file in some way it returns an empty dictionary. This will result in a `KeyError` if you attempt to access a specific key in the empty dictionary. There should either be a check if `names_obj` is empty or the attempt to access a key that may or may not exist should be done in a safe manner (or both).

Commit:
Add check for load_customers

cduhn17 commented:
@

> > > > > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > > > > 
> > > > > 
> > > > > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> > > > 
> > > > 
> > > > @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> > > > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332
> > > 
> > > 
> > > > > > As a note to all involved, the PRs #31 and #33 are extremely old and only valid for our LookingGlass data pulls for which the contract ends on 9/30 and they will be OBE. Is there any possibility we can have them completed before I have to redirect the team to continue other work?
> > > > > 
> > > > > 
> > > > > @mcdonnnj, Although we are going to sunset this version, I was hoping we could at least close out this last pr as it will allow for this version to pass checks. It also leaves us with working scripts in the event we need to revisit this code. Are you able to re-review @cduhn17 's last commit, inputs. Thanks
> > > > 
> > > > 
> > > > @schmelz21 This remains unresolved. As shown below the `names_obj` dictionary is _still_ accessed without any consideration for whether or not `load_customer()` has returned an empty dictionary.
> > > > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L328-L332
> > > 
> > > 
> > > @mcdonnnj, appreciate the quick response and apologies for the miss... circling in @cduhn17 to review and will get back with updates.
> > 
> > 
> > @schmelz21 , @mcdonnnj Per commit [eeb47ed](https://github.com/cisagov/pe-reports/pull/33/commits/eeb47ed69c0e211942434dde817043e0be412267) I was thinking we had addressed the question of if the load_customer return value is empty or not right in the load_customer function.
> > https://github.com/cisagov/pe-reports/blob/7115dd4ff176abfaf8144065cd03f5b35d74ff36/src/pe_reports/report_generator.py#L65-L73
> 
> @cduhn17 All that has been done is making sure that the function returns a dictionary no matter what. However if there is an issue accessing the file in some way it returns an empty dictionary. This will result in a `KeyError` if you attempt to access a specific key in the empty dictionary. There should either be a check if `names_obj` is empty or the attempt to access a key that may or may not exist should be done in a safe manner (or both).

@mcdonnnj ,

I have addressed your recent request for a change to address an empty dictionary see commit [e1df499](https://github.com/cisagov/pe-reports/pull/33/commits/e1df49973566db14e537a42ff38fd1aecb29b05c) https://github.com/cisagov/pe-reports/blob/e1df49973566db14e537a42ff38fd1aecb29b05c/src/pe_reports/report_generator.py#L324-L357

Commit:
Remove exec_info and change return value

Commit:
Change mkdir to makedirs

schmelz21 commented:
Closing pull-request as efforts to update the main branch have been canceled. Routines used in version 0.1.1 have been sunsetted with no future plans to update.
